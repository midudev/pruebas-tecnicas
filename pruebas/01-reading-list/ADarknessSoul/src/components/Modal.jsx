import { useDispatch } from "react-redux";
import { setIsModal } from "../store/slices/WhatABook";

export const Modal = ({
  ISBN = 0,
  author = {},
  cover = "",
  genre = "",
  pages = 0,
  synopsis = "",
  title = "",
  year = 0,
}) => {
  const dispatch = useDispatch();

  return (
    <>
      <div
        className="modal fade"
        data-bs-backdrop="static"
        data-bs-keyboard="false"
        tabIndex="-1"
        aria-labelledby="staticBackdropLabel"
        aria-hidden="true"
        id={`staticBackdrop-${ISBN}`}
      >
        <div className="modal-dialog" style={{ maxWidth: "600px" }}>
          <div className="modal-content">
            <div className="modal-header">
              <h4 className="modal-title">{title}</h4>
              <button
                type="button"
                className="btn-close"
                data-bs-dismiss="modal"
                onClick={() => dispatch(setIsModal(false))}
              ></button>
            </div>

            <div className="modal-body d-flex flex-sm-row flex-column justify-content-center align-items-center">
              <img
                src={cover}
                alt="Portada del libro"
                className="Modal__image"
                height="500px"
                width="300px"
              />

              <ul className="list-group ms-3">
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">Sinopsis:</span> {synopsis}
                </li>
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">Páginas:</span> {pages}
                </li>
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">Género:</span> {genre}
                </li>
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">Año:</span> {year}
                </li>
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">ISBN:</span> {ISBN}
                </li>
                <li className="list-group-item Modal__item">
                  <span className="text-secondColor">Nombre del autor:</span>{" "}
                  {author?.name}
                </li>
                {author?.otherBooks?.length === 0 ? (
                  <></>
                ) : (
                  author?.otherBooks?.map((book) => (
                    <li key={book} className="list-group-item Modal__item">
                      <span className="text-secondColor">Otra obra:</span>{" "}
                      {book}
                    </li>
                  ))
                )}
              </ul>
            </div>

            <div className="modal-footer">
              <button
                type="button"
                className="btn btn-danger"
                data-bs-dismiss="modal"
                onClick={() => dispatch(setIsModal(false))}
                style={{ fontSize: "1.2rem" }}
              >
                Close
              </button>
            </div>
          </div>
        </div>
      </div>
    </>
  );
};
