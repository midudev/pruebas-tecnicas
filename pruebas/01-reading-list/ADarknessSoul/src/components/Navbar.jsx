import { useDispatch, useSelector } from "react-redux";
import {
  allowSave,
  denySave,
  getLocalStorage,
  resetCurrentBook,
  searchBook,
} from "../store/slices/WhatABook";
import { useState } from "react";

export const Navbar = () => {
  const dispatch = useDispatch();
  const { books } = useSelector((state) => state.WhatABook);
  const [inputBook, setInputBook] = useState("");
  const [undo, setUndo] = useState(false); //Deshace los filtros

  const submit = (e) => {
    e.preventDefault();

    setUndo(true);
    if (inputBook) {
      //Se filtra sobre el estado guardado en localStorage. Se deshabilita el guardado para todas las acciones, se resetea el contador de mis libros y favs, finalmente se busca el libro
      dispatch(getLocalStorage());
      dispatch(denySave());
      dispatch(resetCurrentBook());
      dispatch(searchBook(inputBook));
    }
  };

  const onUndo = () => {
    //Deshacer el filtro. Se vuelve al estado previo guardado en localStorage y se habilita el guardado para todas las acciones
    dispatch(getLocalStorage());
    dispatch(allowSave());
    setUndo(false);
  };

  return (
    <>
      <nav className="navbar navbar-expand-lg py-3 Navbar border-bottom align-items-center justify-content-center">
        <a className="navbar-brand mx-5 navbar__logo border" href="/">
          WhatA<span className="fw-bold">Book</span>
        </a>
        {/* navbar__container */}
        <form
          className="form-inline my-2 my-lg-0 mx-auto navbar__form"
          onSubmit={submit}
        >
          <label htmlFor="SearchBook" className="text-center navbar__label">
            Busca un libro
          </label>

          <div className="d-flex flex-sm-row flex-column gap-3">
            <button
              className={
                undo
                  ? "btn btn-danger me-3 d-block"
                  : "btn btn-danger me-3 d-none"
              }
              onClick={onUndo}
              type="button"
            >
              <p className="m-0">Regresar</p>
            </button>
            <input
              className="form-control mr-sm-2 navbar__input"
              type="search"
              placeholder="¿Qué buscas hoy?"
              id="SearchBook"
              onChange={(e) => setInputBook(e.target.value)}
              list="bookNames"
            />
            <datalist id="bookNames">
              {books.map(({ book }) => (
                <option key={book.title}>{book.title}</option>
              ))}
            </datalist>
            <button
              className="btn mx-2 my-sm-0 btn__submit"
              type="submit"
              onClick={submit}
            >
              Buscar
            </button>
          </div>
        </form>
        {/* navbar__form */}
      </nav>
      {/* Navbar */}
    </>
  );
};
