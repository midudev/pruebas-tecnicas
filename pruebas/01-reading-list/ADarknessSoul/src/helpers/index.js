export * from "./getGenres";
export * from "./getNumPages";