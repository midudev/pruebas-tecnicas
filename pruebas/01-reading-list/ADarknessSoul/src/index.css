@font-face {
  font-family: "Montserrat-Regular";
  src: url("./assets/fonts/Montserrat-Regular.ttf") format("truetype");
}

@font-face {
  font-family: "Montserrat-Bold";
  src: url("./assets/fonts/Montserrat-Bold.ttf") format("truetype");
}

:root {
  --main-color: rgb(255, 255, 255);
  --second-color: #5D4737;
  --backgrounds: #F9F3EE;
  --gray: #484646;
  --light-gray: #969090;
  /* --accent: #4F3624; */
  --accent: #fb8500;
  --accent-hover: #EA9432;
  --main-font: "Montserrat-Regular", sans-serif;
  --main-font-bold: "Montserrat-Bold", sans-serif;
}

/*Global*/
html {
  font-size: 62.5%;
  box-sizing: border-box;
}

*,
*:before,
*:after {
  box-sizing: inherit;
}

body {
  background-color: var(--main-color);
  font-size: 1.6rem;
  line-height: 2;
  font-family: var(--main-font);
}

h1,
h2,
h3,
h4,
h5,
h6 {
  color: var(--second-color) !important;
  font-family: var(--main-font-bold);
  line-height: 1.2;
}

h1 {
    font-size: 4.8rem;
}

h2 {
    font-size: 4rem;
}

h3 {
    font-size: 3.2rem;
}

h4 {
    font-size: 2.8rem;
}

img {
    max-width: 100%;
}

a {
    text-decoration: none;
}

label {

  color: var(--second-color);

}

p {
  font-family: var(--main-font);
  font-size: 1.6rem;
}

main {

  width: 95%;

}

/* Utilities */

.no-filter {

  filter: none;

}

.text-secondColor {

  color: var(--second-color);
  font-weight: bold;

}

/* Navbar */

.Navbar {

  background-color: var(--backgrounds);

}

@media (min-width: 350px) {
    
  .navbar__container {

    margin-left: 1rem;
    margin-top: 0;
    margin-bottom: 0;

  }

}

.navbar__logo {

  font-size: 1.6rem;
  background-color: var(--main-color);
  padding: 1.6rem;
  border-radius: 0.5rem;
  margin-top: 2rem;
  margin-bottom: 1rem;

}

@media (min-width: 418px) {
    .navbar__logo {

      margin-top: 1rem;
      margin-bottom: 1rem;

    }
}

.btn__submit {

  background-color: var(--accent);
  color: white;
  font-size: 1.2rem;

}

.btn__submit:hover {

  background-color: var(--accent-hover);
  color: white;

}

.navbar__label {

  font-size: 2rem;
  color: var(--second-color);

}

.navbar__form {

  margin-bottom: 2rem!important;

}

@media (min-width: 418px) {
    
  .navbar__form {

    margin-bottom: 1rem!important;

  }

}

.navbar__input {

  font-size: 1.2rem;

}

/* Filters */

.btn__filter {

  background-color: var(--accent);
  color: var(--main-color);

}

.btn__filter:hover {

  background-color: var(--accent-hover);
  color: var(--main-color);

}

.filter__select {

  font-size: 1.2rem;

}

.form__Input {

  background-color: var(--accent)!important;
  color: var(--main-color);
  font-size: 1.2rem;

}

.form__Input:hover {

  background-color: var(--accent-hover)!important;
  color: var(--main-color)!important;

}

/* Books */

.Books {

  display: flex;
  flex-wrap: nowrap;
  position: relative;
  justify-content: center;
  flex-direction: column-reverse;
  align-items: center;

}

@media (min-width: 1024px) {
    
  .Books {

    flex-direction: row;
    align-items: flex-start;

  }

}

.Books__available {

  background-color: var(--backgrounds);
  box-shadow: -5px 7px 5px -4px rgba(72,70,70,0.31);
  flex: 0 0 66%;

}

.Books__myBooks {

  display: flex;
  flex-direction: column;
  margin-bottom: 1rem;
  height: 97vh;
  max-width: min(35rem, 100%);
  background-color: var(--backgrounds);
  box-shadow: -5px 7px 5px -4px rgba(72,70,70,0.31);
  flex: 1;
  position: static;
  top: 1rem;

}

.flex-grow-1 {
  flex-grow: 1; /* Make this div (BookCard container) take up all the available space */
}

@media (min-width: 1024px) {
  
.Books__myBooks {

  margin-left: 1rem;
  margin-bottom: 0;

}


}

.Books__header {

  text-align: center;
  margin: 1rem 1rem;
  background-color: var(--main-color);
  padding: 0.5rem;
  border-radius: 0.5rem;

}

.Books__cards {

  display: grid;
  grid-template-columns: 1fr;
  gap: 1rem;

}

.Books__cards--smaller {

    display: grid;
    grid-template-columns: 1fr;
    gap: 1rem;

}

@media (min-width: 480px) {
    
  .Books__cards {

    grid-template-columns: repeat(2, 1fr)!important;

  }

  
  .Books__cards--smaller {

    grid-template-columns: repeat(2, 1fr)!important;

  }

}

@media (min-width: 768px) {
    
  .Books__cards {

    grid-template-columns: repeat(3, 1fr)!important;

  }

  .Books__cards--smaller {

    grid-template-columns: repeat(3, 1fr)!important;

  }

}

@media (min-width: 1024px) {
    
  .Books__cards {

    grid-template-columns: repeat(4, 1fr)!important;

  }

}

@media (min-width: 1200px) {
    
  .Books__cards {

    grid-template-columns: repeat(5, 1fr)!important;

  }

}

/* BookCard */

.Book__card {

  height: 35rem;

}

.Book__card--size {

  height: 100%;

}

.Book__figure {
  filter: brightness(40%);
  transition: all 0.5s ease;
}

.Book__figure:hover {
  filter: brightness(100%);
  cursor: pointer;
}

.Img__figcaption {
  top: 50%;
  left: 50%;
  margin-right: -50%;
  transform: translate(-50%, -50%);
  cursor: pointer;
  z-index: 1;
  color: var(--main-color);
  background-color: rgba(47, 47, 47, 0.606);
  padding: 0.5rem;
  border-radius: 1rem;
  visibility: hidden;
  
}

.Img__figcaption--numbers {

  /* bottom: 7px;
  right: 7px; */
  top: 7px;
  right: 7px;
  color: var(--main-color);
  background-color: rgba(47, 47, 47, 0.606);
  padding: 0.5rem;
  border-radius: 1rem;
  
}

.Book__figure:hover .Img__figcaption {
  visibility: visible;
  transition: visibility 0.5s ease;
}


.Book__title {

  font-size: 1.6rem;
  text-align: center;

}

.Book__synopsis {

  font-size: 1.3rem;
  color: var(--second-color)

}

.Book__agregar {

  font-size: 1.2rem;
  background-color: var(--accent);
  color: var(--main-color);
  border-start-start-radius: 0;
  border-start-end-radius: 0;
  padding: 0.8rem;

}

.Book__agregar:hover {

  background-color: var(--accent-hover);
  color: var(--main-color);

}

.Book__invert {

  filter: invert(100%);

}

.Book__buttonIcon {

  flex: 0 0 calc(50%);
  border-radius: 0;
  padding: 0.5rem;
  

}

.Book__footerIcons:first-child {

  background-color: var(--accent);
  flex: 0 0 calc(50% - 1px);
  border-start-start-radius: 0;
  border-end-end-radius: 0;
    padding: 0.5rem;

}

.Book__footerIcons:last-child {

  background-color: var(--accent);
  flex: 0 0 calc(50% - 1px);
  border-start-end-radius: 0;
  border-end-start-radius: 0;
  padding: 0.5rem;
}

.Book__footerIcons:hover {

  background-color: var(--accent-hover);

}

.Book__footerIcons img {

  filter: invert(100%);

}

.Books__headerButtons:first-child {

  flex: 0 0 calc(50% - 1px);
  font-size: 1.5rem;
  background-color: var(--accent);
  color: var(--main-color);
  border-end-start-radius: 0;
  border-start-end-radius: 0;

}


.Books__headerButtons:last-child {

  flex: 0 0 calc(50% - 1px);
  font-size: 1.5rem;
  background-color: var(--accent);
  color: var(--main-color);
  border-start-start-radius: 0;
  border-end-end-radius: 0;

}

.Books__headerButtons:hover {

  background-color: var(--accent-hover);

}

/* Modal */

.Modal__item {

  font-size: 1.4rem;

}

.Modal__image {

  height: inherit;
  margin-bottom: 1rem;

}

@media (min-width: 580px) {
    
  .Modal__image {

    margin-bottom: 0;

  }

}

.Search__button {

  filter: invert(100%);

}