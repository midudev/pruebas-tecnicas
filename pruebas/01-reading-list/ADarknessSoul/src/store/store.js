import { configureStore } from '@reduxjs/toolkit';
import { WhatABookSlice } from './slices/WhatABook';

export const store = configureStore({
  reducer: {

    WhatABook: WhatABookSlice.reducer,

  },
});