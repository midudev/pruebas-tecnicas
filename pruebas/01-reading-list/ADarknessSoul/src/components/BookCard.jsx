import { useDispatch, useSelector } from "react-redux";
import {
  pushMyBook,
  removeAvailableBook,
  pushAvailableBooks,
  removeMyBook,
  pushMyFavs,
  removeFavorite,
  setIsModal,
  saveLocalStorage,
} from "../store/slices/WhatABook/";
import { Modal } from "./Modal";
import { useEffect } from "react";

export const BookCard = ({ book = {}, type, isFavorite = false }) => {
  const dispatch = useDispatch();
  const { myBooks, currentBook, myFavs, booksAvailable, isModal } = useSelector(
    (state) => state.WhatABook
  );

  //Cada useEffect guarda el estado en localStorage cuando se hace un movimiento en un valor del store
  useEffect(() => {
    dispatch(saveLocalStorage());
  }, [myBooks]);

  useEffect(() => {
    dispatch(saveLocalStorage());
  }, [booksAvailable]);

  useEffect(() => {
    dispatch(saveLocalStorage());
  }, [myFavs]);

  useEffect(() => {
    dispatch(saveLocalStorage());
  }, [currentBook]);

  useEffect(() => {
    dispatch(saveLocalStorage());
  }, [isModal]);

  const onMoveBook = () => {
    //Agregar el libro al array de mis libros, y lo quita de los libros disponibles
    dispatch(pushMyBook(book));
    dispatch(removeAvailableBook(book));
  };

  const onRemoveBook = () => {
    //Agregar el libro al array de libros disponibles, y lo quita de mis libros, y favoritos
    dispatch(pushAvailableBooks(book));
    dispatch(removeMyBook(book));
    dispatch(removeFavorite(book));
  };

  const onRemoveFavorite = () => {
    dispatch(removeFavorite(book));
  };

  const onFavoriteBook = () => {
    dispatch(pushMyFavs(book));
  };

  return (
    <>
      <div className={type === 1 ? "card p-0" : "card p-0 Book__card--size"}>
        <figure className="position-relative Book__figure">
          <img
            src={book?.cover}
            alt="Portada del libro"
            width="100%"
            className="Book__card rounded-top"
            data-bs-toggle="modal"
            data-bs-target={`#staticBackdrop-${book.ISBN}`}
            onClick={() => dispatch(setIsModal(true))}
          />
          <figcaption
            className="position-absolute Img__figcaption"
            data-bs-toggle="modal"
            data-bs-target={`#staticBackdrop-${book.ISBN}`}
            onClick={() => dispatch(setIsModal(true))}
          >
            Ver m치s...
          </figcaption>
        </figure>

        {type === 1 ? (
          <></>
        ) : (
          <div className="position-absolute Img__figcaption--numbers no-filter">
            {currentBook + 1}/
            {isFavorite ? `${myFavs.length}` : `${myBooks.length}`}
          </div>
        )}

        <div className="card-body d-flex flex-column p-0">
          <div className="px-3 pt-3">
            <h2 className="Book__title mt-3 mb-3">{book?.title}</h2>

            <hr />

            <p className="Book__synopsis">{book?.synopsis}</p>
          </div>

          {type === 1 ? (
            <div className="mt-auto">
              <button
                className="btn col-12 Book__agregar mt-3"
                onClick={onMoveBook}
              >
                Agregar
              </button>
            </div>
          ) : !isFavorite ? (
            <div className="mt-auto d-flex justify-content-around">
              <button
                className="btn btn-success Book__buttonIcon"
                onClick={onFavoriteBook}
              >
                <img
                  src="/src/assets/images/hearth.svg"
                  alt="Bot칩n de me gusta"
                  width="30px"
                  height="30px"
                  className="Book__invert"
                />
              </button>
              <button
                className="btn btn-danger Book__buttonIcon"
                onClick={onRemoveBook}
              >
                <img
                  src="/src/assets/images/delete.svg"
                  alt="Bot칩n de borrar"
                  width="30px"
                  height="30px"
                  className="Book__invert"
                />
              </button>
            </div>
          ) : (
            <div className="mt-auto">
              <button
                className="btn btn-danger Book__buttonIcon col-12"
                onClick={onRemoveFavorite}
              >
                <img
                  src="/src/assets/images/delete.svg"
                  alt="Bot칩n de borrar"
                  width="30px"
                  height="30px"
                  className="Book__invert"
                />
              </button>
            </div>
          )}
        </div>
      </div>

      <Modal {...book} />
    </>
  );
};
