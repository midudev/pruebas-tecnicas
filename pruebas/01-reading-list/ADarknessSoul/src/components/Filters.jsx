import { useDispatch, useSelector } from "react-redux";
import { getAllGenres, getMaxPages, getMinPages } from "../helpers/";
import { useEffect, useState } from "react";
import {
  allowSave,
  denySave,
  filterGenre,
  filterRange,
  getLocalStorage,
  resetCurrentBook,
} from "../store/slices/WhatABook";

export const Filters = () => {
  const dispatch = useDispatch();
  const { books = [] } = useSelector((state) => state.WhatABook);
  const [minNumPages, setMinNumPages] = useState(0);
  const [maxNumPages, setMaxNumPages] = useState(0);
  const [rangeNumber, setRangeNumber] = useState(0);
  const [currentGenre, setCurrentGenre] = useState("");
  const [genres, setGenres] = useState([]);
  const [undo, setUndo] = useState(false); //Dehacer filtros

  useEffect(() => {
    //Se obtiene el mínimo y máximo número de páginas, así como los generos, y se guardan en los states
    const min = getMinPages(books);
    setMaxNumPages(getMaxPages(books));
    setGenres(getAllGenres(books));
    setMinNumPages(min);
    setRangeNumber(min);
  }, [books]);

  const onRangeChange = (e) => {
    setRangeNumber(Number(e.target.value));
  };

  const onSubmitRange = (e) => {
    //Se obtiene localStorage para filtrar desde él, se deshabilita el guardado para las demás acciones, se resetea el contador de mis libros y se aplica el filtro
    e.preventDefault();
    setUndo(true);
    dispatch(getLocalStorage());
    dispatch(denySave());
    dispatch(resetCurrentBook());
    dispatch(filterRange(rangeNumber));
  };

  const onSubmitGenre = (e) => {
    //Se obtiene localStorage para filtrar desde él, se deshabilita el guardado para las demás acciones, se resetea el contador de mis libros y se aplica el filtro
    e.preventDefault();
    if (currentGenre !== "default") {
      setUndo(true);
      dispatch(getLocalStorage());
      dispatch(denySave());
      dispatch(resetCurrentBook());
      dispatch(filterGenre(currentGenre));
    }
  };

  const onDecrementRange = () => {
    //Decrementa el state y el value del input
    if (rangeNumber === minNumPages) return;
    setRangeNumber(rangeNumber - 1);
    const range = document.getElementById("NumPaginas");
    range.value = rangeNumber - 1;
  };

  const onIncrementRange = () => {
    //Decrementa el state y el value del input
    if (rangeNumber === maxNumPages) return;
    setRangeNumber(rangeNumber + 1);
    const range = document.getElementById("NumPaginas");
    range.value = rangeNumber + 1;
  };

  const onUndo = () => {
    //Deshacer el filtro. Se vuelve al estado previo guardado en localStorage y se habilita el guardado para todas las acciones
    dispatch(getLocalStorage());
    dispatch(allowSave());
    setUndo(false);
  };

  return (
    <>
      <section className="container mt-3">
        <div className="row gap-3">
          <form
            className="d-flex flex-column col-sm"
            onSubmit={(e) => onSubmitRange(e)}
          >
            <label className="mb-3">Buscar por número de páginas</label>
            <input
              type="range"
              min={minNumPages}
              max={maxNumPages}
              id="NumPaginas"
              onChange={(e) => onRangeChange(e)}
              value={rangeNumber}
            />

            <div className="d-flex justify-content-center gap-3 align-items-center mt-2 mb-3">
              <button
                className="btn form_buttonLeft"
                onClick={onDecrementRange}
                type="button"
              >
                <img
                  src="/src/assets/images/leftArrow.svg"
                  alt="Flecha izquierda"
                  width="20px"
                />
              </button>
              <p className="text-center m-0">{rangeNumber}</p>
              <button
                className="btn form_buttonRight"
                onClick={onIncrementRange}
                type="button"
              >
                <img
                  src="/src/assets/images/rightArrow.svg"
                  alt="Flecha izquierda"
                  width="20px"
                />
              </button>
            </div>

            <div className="d-flex justify-content-center mt-auto">
              <input
                type="submit"
                className="btn col-md-3 col-12 form__Input"
                value="Buscar"
              />
            </div>
          </form> {/* form1 */}

          <form
            className="col-sm d-flex flex-column"
            onSubmit={(e) => onSubmitGenre(e)}
          >
            <label className="mb-3">Buscar por genero</label>

            <select
              name="genre"
              id="genre"
              className="form-select filter__select mb-3"
              onChange={(e) => setCurrentGenre(e.target.value)}
            >
              <option value="default">-</option>

              {genres.map((genre) => (
                <option key={genre} value={genre}>
                  {genre}
                </option>
              ))}
            </select>

            <div className="d-flex justify-content-center mt-auto">
              <input
                type="submit"
                className="btn col-md-3 col-12 form__Input"
                value="Buscar"
              />
            </div>
          </form> {/* form2 */}
        </div>

        <div className="d-flex justify-content-center mt-3">
          <button
            className={
              !undo ? "btn btn-danger d-none" : "btn btn-danger d-block"
            }
            style={{ fontSize: "1.3rem" }}
            onClick={onUndo}
          >
            Deshacer
          </button>
        </div>
      </section>
    </>
  );
};
