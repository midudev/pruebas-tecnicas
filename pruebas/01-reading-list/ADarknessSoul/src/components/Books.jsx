import { useDispatch, useSelector } from "react-redux";
import { BookCard } from "./BookCard";
import {
  incrementCurrentBook,
  decrementCurrentBook,
  resetCurrentBook,
  setIsFav,
  setAll,
  saveLocalStorage,
} from "../store/slices/WhatABook/";
import { useEffect, useState } from "react";

export const Books = () => {
  const dispatch = useDispatch();
  const { booksAvailable, myBooks, currentBook, myFavs, isFav, isModal } =
    useSelector((state) => state.WhatABook);
  const [reduxBooksAvailable, setReduxBooksAvailable] = useState([]);
  const [reduxMyFavs, setReduxMyFavs] = useState([]);
  const [reduxMyBooks, setReduxMyBooks] = useState([]);
  const [screenWidth, setScreenWidth] = useState(window.innerWidth); //Se guarda la resolución actual para arreglar un problema de el sidebar con mis libros y favoritos

  useEffect(() => {
    //Se guardan los estados de redux en un state para asegurar el correcto renderizado
    setReduxBooksAvailable([...booksAvailable]);
    setReduxMyBooks([...myBooks]);
    setReduxMyFavs([...myFavs]);
  }, [booksAvailable, myBooks, myFavs]);

  useEffect(() => {
    //Se guardan los cambios en localStorage cada que cambia el array de favoritos en el store
    dispatch(saveLocalStorage());
  }, [isFav]);

  const handleWindowChange = () => {
    setScreenWidth(window.innerWidth);
  };

  useEffect(() => {
    //Se guarda la resolución cada que cambia el tamaño de pantalla
    window.addEventListener("resize", handleWindowChange);

    return () => {
      window.removeEventListener("resize", handleWindowChange);
    };
  }, []);

  return (
    <>
      <main className="mt-5 mx-auto mb-3 h-100">
        <div className={myBooks.length === 0 ? "" : "Books"}>
          <section className="Books__available rounded border">
            <div className="d-flex flex-column Books__header">
              <h4>
                Libros disponibles: <span>{booksAvailable.length}</span>
              </h4>
            </div>

            <div
              className={
                myBooks.length === 0
                  ? "mx-3 Books__cards mb-3"
                  : "mx-3 Books__cards--smaller mb-3"
              }
            >
              {reduxBooksAvailable.map(({ book }, index) => (
                <BookCard key={index} book={book} type={1} />
              ))}
            </div>
          </section>

          {myBooks.length === 0 ? (
            <></>
          ) : (
            <section
              className={
                screenWidth <= 1024
                  ? "Books__myBooks rounded border position-static"
                  : isModal
                  ? "Books__myBooks rounded border position-static"
                  : "Books__myBooks rounded border position-sticky"
              }
            >
              <div className="d-flex justify-content-between">
                <button
                  className="btn Books__headerButtons"
                  onClick={() => {
                    dispatch(setAll());
                    dispatch(resetCurrentBook());
                  }}
                >
                  Todos
                </button>
                <button
                  className="btn Books__headerButtons"
                  onClick={() => {
                    dispatch(setIsFav());
                    dispatch(resetCurrentBook());
                  }}
                >
                  Favoritos
                </button>
              </div>

              <div className="d-flex flex-column Books__header">
                <h4>
                  {isFav ? "Mis favoritos: " : "Mis libros: "}
                  <span>
                    {!isFav ? `${myBooks.length}` : `${myFavs.length}`}
                  </span>
                </h4>
              </div>

              <div
                className={
                    "mx-3 flex-grow-1 d-flex flex-column justify-content-center align-items-center"
                }
              >
                {reduxMyBooks.length === 0 ? (
                  <></>
                ) : !isFav ? (
                  <BookCard book={reduxMyBooks[currentBook]?.book} type={2} />
                ) : myFavs.length === 0 ? (
                    <h4 className="text-center">Aun no tienes favoritos</h4>
                ) : (
                  <BookCard
                    book={reduxMyFavs[currentBook]?.book}
                    type={2}
                    isFavorite={isFav}
                  />
                )}
              </div>

              <div className="d-flex justify-content-between mt-5 Book__footer">
                <button
                  className="btn Book__footerIcons"
                  onClick={() => dispatch(decrementCurrentBook())}
                >
                  <img
                    src="/src/assets/images/leftArrow.svg"
                    alt="Flecha izquierda"
                    width="20px"
                    height="20px"
                  />
                </button>
                <button
                  className="btn Book__footerIcons"
                  onClick={() => dispatch(incrementCurrentBook())}
                >
                  <img
                    src="/src/assets/images/rightArrow.svg"
                    alt="Flecha derecha"
                    width="20px"
                    height="20px"
                  />
                </button>
              </div>
            </section>
          )}
        </div>
      </main>
    </>
  );
};
