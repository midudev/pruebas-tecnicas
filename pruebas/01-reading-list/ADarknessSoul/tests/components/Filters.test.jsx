import { Provider } from "react-redux";
import { store } from "../../src/store";
import { render } from "@testing-library/react";
import { Filters } from "../../src/components/Filters";

describe('Pruebas en <Books/>', ()=> {

    test('Debe hacer match con el snapshot', ()=> {

        const component = render(

            <Provider store={store}>

                <Filters/>

            </Provider>

        );

        expect(component).toMatchSnapshot();

    }); 

});