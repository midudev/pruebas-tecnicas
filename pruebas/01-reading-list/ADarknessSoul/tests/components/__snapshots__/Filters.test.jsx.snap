// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Books/> Debe hacer match con el snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="container mt-3"
      >
        <div
          class="row gap-3"
        >
          <form
            class="d-flex flex-column col-sm"
          >
            <label
              class="mb-3"
            >
              Buscar por número de páginas
            </label>
            <input
              id="NumPaginas"
              max="0"
              type="range"
              value="0"
            />
            <div
              class="d-flex justify-content-center gap-3 align-items-center mt-2 mb-3"
            >
              <button
                class="btn form_buttonLeft"
                type="button"
              >
                <img
                  alt="Flecha izquierda"
                  src="/src/assets/images/leftArrow.svg"
                  width="20px"
                />
              </button>
              <p
                class="text-center m-0"
              />
              <button
                class="btn form_buttonRight"
                type="button"
              >
                <img
                  alt="Flecha izquierda"
                  src="/src/assets/images/rightArrow.svg"
                  width="20px"
                />
              </button>
            </div>
            <div
              class="d-flex justify-content-center mt-auto"
            >
              <input
                class="btn col-md-3 col-12 form__Input"
                type="submit"
                value="Buscar"
              />
            </div>
          </form>
           
          <form
            class="col-sm d-flex flex-column"
          >
            <label
              class="mb-3"
            >
              Buscar por genero
            </label>
            <select
              class="form-select filter__select mb-3"
              id="genre"
              name="genre"
            >
              <option
                value="default"
              >
                -
              </option>
            </select>
            <div
              class="d-flex justify-content-center mt-auto"
            >
              <input
                class="btn col-md-3 col-12 form__Input"
                type="submit"
                value="Buscar"
              />
            </div>
          </form>
           
        </div>
        <div
          class="d-flex justify-content-center mt-3"
        >
          <button
            class="btn btn-danger d-none"
            style="font-size: 1.3rem;"
          >
            Deshacer
          </button>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="container mt-3"
    >
      <div
        class="row gap-3"
      >
        <form
          class="d-flex flex-column col-sm"
        >
          <label
            class="mb-3"
          >
            Buscar por número de páginas
          </label>
          <input
            id="NumPaginas"
            max="0"
            type="range"
            value="0"
          />
          <div
            class="d-flex justify-content-center gap-3 align-items-center mt-2 mb-3"
          >
            <button
              class="btn form_buttonLeft"
              type="button"
            >
              <img
                alt="Flecha izquierda"
                src="/src/assets/images/leftArrow.svg"
                width="20px"
              />
            </button>
            <p
              class="text-center m-0"
            />
            <button
              class="btn form_buttonRight"
              type="button"
            >
              <img
                alt="Flecha izquierda"
                src="/src/assets/images/rightArrow.svg"
                width="20px"
              />
            </button>
          </div>
          <div
            class="d-flex justify-content-center mt-auto"
          >
            <input
              class="btn col-md-3 col-12 form__Input"
              type="submit"
              value="Buscar"
            />
          </div>
        </form>
         
        <form
          class="col-sm d-flex flex-column"
        >
          <label
            class="mb-3"
          >
            Buscar por genero
          </label>
          <select
            class="form-select filter__select mb-3"
            id="genre"
            name="genre"
          >
            <option
              value="default"
            >
              -
            </option>
          </select>
          <div
            class="d-flex justify-content-center mt-auto"
          >
            <input
              class="btn col-md-3 col-12 form__Input"
              type="submit"
              value="Buscar"
            />
          </div>
        </form>
         
      </div>
      <div
        class="d-flex justify-content-center mt-3"
      >
        <button
          class="btn btn-danger d-none"
          style="font-size: 1.3rem;"
        >
          Deshacer
        </button>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
